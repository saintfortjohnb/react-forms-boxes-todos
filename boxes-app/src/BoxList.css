/* src/BoxList.css */
.BoxList {
    display: flex;
    flex-wrap: wrap;
    gap: 10px; 
    margin-top: 20px;
}
