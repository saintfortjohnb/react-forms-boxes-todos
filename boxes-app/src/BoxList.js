// src/BoxList.js
import React, { useState } from 'react';
import Box from './Box';
import NewBoxForm from './NewBoxForm';
import './BoxList.css';

function BoxList() {
    const [boxes, setBoxes] = useState([]);

    const addBox = box => {
        setBoxes(boxes => [...boxes, box]);
    };

    const removeBox = id => {
        setBoxes(boxes => boxes.filter(box => box.id !== id));
    };

    return (
        <div>
            <NewBoxForm addBox={addBox} />
            {boxes.map(box => (
                <Box key={box.id} {...box} removeBox={removeBox} />
            ))}
        </div>
    );
}

export default BoxList;
