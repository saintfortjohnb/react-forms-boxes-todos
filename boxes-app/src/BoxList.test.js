// src/__tests__/BoxList.test.js
import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import BoxList from './BoxList';

it("renders without crashing", function() {
    render(<BoxList />);
});

it("matches snapshot", function() {
    const { asFragment } = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
});

it("can add a new box via the form", function() {
    const { getByPlaceholderText, getByText } = render(<BoxList />);

    // Get form inputs
    const widthInput = getByPlaceholderText("Width");
    const heightInput = getByPlaceholderText("Height");
    const bgColorInput = getByPlaceholderText("Background Color");
    
    // Fire events to simulate user filling out form
    fireEvent.change(widthInput, { target: { value: '100' } });
    fireEvent.change(heightInput, { target: { value: '100' } });
    fireEvent.change(bgColorInput, { target: { value: 'red' } });
    
    const addButton = getByText("Add Box");
    fireEvent.click(addButton);

    // Check that a new box has been added
    expect(getByText('X')).toBeInTheDocument();
});

it("new box appears in the DOM with the correct size and color", function() {
    const { getByPlaceholderText, getByText, container } = render(<BoxList />);

    const widthInput = getByPlaceholderText("Width");
    const heightInput = getByPlaceholderText("Height");
    const bgColorInput = getByPlaceholderText("Background Color");

    fireEvent.change(widthInput, { target: { value: '100' } });
    fireEvent.change(heightInput, { target: { value: '100' } });
    fireEvent.change(bgColorInput, { target: { value: 'red' } });

    const addButton = getByText("Add Box");
    fireEvent.click(addButton);

    const addedBox = container.querySelector("div[style='width: 100px; height: 100px; background-color: red;']");
    expect(addedBox).toBeInTheDocument();
});

it("can remove a box", function() {
    const { getByPlaceholderText, getByText, queryByText, container } = render(<BoxList />);

    const widthInput = getByPlaceholderText("Width");
    const heightInput = getByPlaceholderText("Height");
    const bgColorInput = getByPlaceholderText("Background Color");

    fireEvent.change(widthInput, { target: { value: '100' } });
    fireEvent.change(heightInput, { target: { value: '100' } });
    fireEvent.change(bgColorInput, { target: { value: 'red' } });

    const addButton = getByText("Add Box");
    fireEvent.click(addButton);

    // Now, we have added the box. Let's remove it.
    const removeButton = getByText('X');
    fireEvent.click(removeButton);

    // Ensure the box and the 'X' button are no longer in the document
    const removedBox = container.querySelector("div[style='width: 100px; height: 100px; background-color: red;']");
    expect(removedBox).not.toBeInTheDocument();
    expect(queryByText('X')).not.toBeInTheDocument();
});
