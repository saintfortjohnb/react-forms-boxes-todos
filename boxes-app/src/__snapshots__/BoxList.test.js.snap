// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <form>
      <input
        name="width"
        placeholder="Width"
        value=""
      />
      <input
        name="height"
        placeholder="Height"
        value=""
      />
      <input
        name="backgroundColor"
        placeholder="Background Color"
        value=""
      />
      <button>
        Add Box
      </button>
    </form>
  </div>
</DocumentFragment>
`;
